### ISSUE TITLE
[refmt] hangs & pegs CPU for a certain ocaml file

### ISSUE THREAD

**Issue by jaredly** [facebook/reason#2190]
Here's the file https://gist.github.com/jaredly/5a6eb7c66e5919ac932a053271785d8c
this is version 3.3.3.
No idea what's going on.

**Comment by anmonteiro** [comment]
I'm looking at this

**Comment by anmonteiro** [comment]
Turns out this is not an infinite loop, but the function here https://github.com/facebook/reason/blob/08ff059f8504a8a60a9ae486a0e469137e1f7588/src/reason-parser/reason_pprint_ast.ml#L1774
takes a very long time to execute.

**Comment by anmonteiro** [comment]
Here's a smaller version of the Gist that is enough to repro the problem:

<details>
<summary>Click to expand</summary>

```ocaml
let deserialize_Data__Node__t contentsTransformer prefixTransformer record =
  match Js.Json.classify record with
  | ((JSONObject (dict))[@explicit_arity ]) ->
    (match Js.Dict.get dict "prefix" with
     | ((Some (json))[@explicit_arity ]) ->
       (match prefixTransformer json with
        | ((Ok (attr_prefix))[@explicit_arity ]) ->
          (match Js.Dict.get dict "tags" with
           | ((Some (json))[@explicit_arity ]) ->
             (match deserialize_Belt_SetString____t json with
              | ((Ok (attr_tags))[@explicit_arity ]) ->
                (match Js.Dict.get dict "contents" with
                 | ((Some (json))[@explicit_arity ]) ->
                   (match contentsTransformer json with
                    | ((Ok
                          (attr_contents))[@explicit_arity ])
                      ->
                      (match Js.Dict.get dict "children"
                       with
                       | ((Some
                             (json))[@explicit_arity ]) ->
                         (match foo
                          with
                          | ((Ok
                                (attr_children))[@explicit_arity
                             ])
                            ->
                            (match Js.Dict.get dict
                                     "childrenModified"
                             with
                             | ((Some
                                   (json))[@explicit_arity
                                ])
                               ->
                               (match deserialize_Data____date
                                        json
                                with
                                | ((Ok
                                      (attr_childrenModified))
                                     [@explicit_arity
                                     ])
                                  ->
                                  (match
                                     Js.Dict.get
                                       dict
                                       "modified"
                                   with
                                   | ((Some
                                         (json))
                                        [@explicit_arity
                                        ])
                                     ->
                                     (match
                                        deserialize_Data____date
                                          json
                                      with
                                      |
                                        ((Ok
                                            (attr_modified))
                                           [@explicit_arity
                                           ]) ->
                                        (match
                                           Js.Dict.get
                                             dict
                                             "trashed"
                                         with
                                         |
                                           ((Some
                                               (json))
                                              [@explicit_arity
                                              ]) ->
                                           () )))))))))))))
```

</details>

**Comment by Lupus** [comment]
I'm running into similar issue with `Reason 3.6.0 @ 59339f6`. I wonder if this included in the 3.6.0 release or I'm seeing a different issue? :thinking:

**Comment by Lupus** [comment]
This is how refmt looks like in htop:
![image](https://user-images.githubusercontent.com/105348/84996090-7adf1c00-b155-11ea-89f9-3cf07a1c01f2.png)
